src/storage/index/mod.rs

```diff
@@ -0,0 +1,529 @@
+// #[allow(warnings)]
+pub mod art;
+pub mod iter;
+pub mod node;
+pub mod snapshot;
+
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::fmt::Debug;
+
+// "Partial" in the Adaptive Radix Tree paper refers to "partial keys", a technique employed
+// for prefix compression in this data structure. Instead of storing entire keys in the nodes,
+// ART nodes often only store partial keys, which are the differing prefixes of the keys.
+// This approach significantly reduces the memory requirements of the data structure.
+// Key is a trait that provides an abstraction for partial keys.
+pub trait Key {
+    fn at(&self, pos: usize) -> u8;
+    fn len(&self) -> usize;
+    fn prefix_before(&self, length: usize) -> Self;
+    fn prefix_after(&self, start: usize) -> Self;
+    fn longest_common_prefix(&self, slice: &[u8]) -> usize;
+    fn as_slice(&self) -> &[u8];
+}
+
+pub trait KeyTrait:
+    Key + Clone + PartialEq + PartialOrd + Ord + Debug + for<'a> From<&'a [u8]>
+{
+}
+impl<T: Key + Clone + PartialOrd + PartialEq + Ord + Debug + for<'a> From<&'a [u8]>> KeyTrait
+    for T
+{
+}
+
+/*
+    Key trait implementations
+*/
+
+// Source: https://www.the-paper-trail.org/post/art-paper-notes/
+//
+// Keys can be of two types:
+// 1. Fixed-length datatypes such as 128-bit integers, or strings of exactly 64-bytes,
+// don’t have any problem because there can, by construction, never be any key that’s
+// a prefix of any other.
+//
+// 2. Variable-length datatypes such as general strings, can be transformed into types
+// where no key is the prefix of any other by a simple trick: append the NULL byte to every key.
+// The NULL byte, as it does in C-style strings, indicates that this is the end of the key, and
+// no characters can come after it. Therefore no string with a null-byte can be a prefix of any other,
+// because no string can have any characters after the NULL byte!
```

Comment:
Just a note that rust strings can freely contain null bytes. `"foo\0bar"` is allowed in rust and will have a lenght of 7. So in rust adding a null byte you might still end up with a prefixed string.

Maybe instead of a null byte you could add something like `0b10111111`? All rust strings are valid UTF-8 and a UTF-8 character can never begin with a `10` bit pattern as this the start pattern of a continuation byte.

Reply:
This is interesting! Thanks for pointing it out @DelSkayn, looking into it

Reply:
Just a suggestion, just ensuring that strings have no internal null bytes is also a valid solution.

Reply:
Thanks @DelSkayn, I am referencing this issue [here](https://github.com/surrealdb/tart/issues/13) and will explore working on it in the library, as we plan to import that as the index eventually.